scratch/copy-mk.py

```diff
@@ -0,0 +1,162 @@
+
+This script makes use of sim_correlator.py in order to start a correlator, but
+instead of a simulated one, it gets the parameters from a live MK correlator.
+"""
+
+import argparse
+import asyncio
+import subprocess
+import sys
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass
+
+import aiokatcp
+
+
+@dataclass
+class Subordinate:
+    name: str
+    control_port: int
+    monitor_port: int
+    digitiser_address: str
+    n_antennas: int
+
+
+def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
+    """Parse command-line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("cmc_host", help="Hostname / IP address of the CMC.")
+    parser.add_argument("--cmc_port", type=int, default=7147, help="KATCP port of the CMC. [%(default)s]")
+    parser.add_argument("--target", type=str, required=True, help="Which subordinate to copy.")
+    parser.add_argument(
+        "--develop",
+        nargs="?",
+        const=True,
+        help="Pass development options in the config. Use comma separation, or omit the arg to enable all.",
+    )
+    parser.add_argument("--image-override", action="append", metavar="NAME:IMAGE:TAG", help="Override a single image")
+    args = parser.parse_args(argv)
+    return args
+
+
+async def get_sensor_value(client: aiokatcp.Client, sensor_name: str, sensor_type: Callable):
+    """Retrieve a sensor value from a katcp client."""
+    try:
+        _, informs = await client.request("sensor-value", sensor_name)
+    except (aiokatcp.FailReply, ConnectionError) as exc:
+        print(f"Error: {exc}", file=sys.stderr)
+        sys.exit(1)
+    return sensor_type(informs[0].arguments[4].decode())
+
+
+async def async_main(args) -> int:
+    cmc_client = await aiokatcp.Client.connect(args.cmc_host, args.cmc_port)
+
+    try:
+        reply, informs = await cmc_client.request("subordinate-list")
+    except (aiokatcp.FailReply, ConnectionError) as exc:
+        print(f"Error: {exc}", file=sys.stderr)
+        return 1
+
+    if len(informs) == 0:
+        print(f"No subarrays currently running on {args.cmc_host}")
+        return 1
+
+    subordinates = {}
+    for inform in informs:
+        name = inform.arguments[0].decode()
+        control_port, monitor_port = inform.arguments[1].decode().split(",")
+        control_port = int(control_port)
+        monitor_port = int(monitor_port)
+        mcast_groups = []
+        for grp in inform.arguments[2:]:
+            mcast_groups.append(grp.decode())
+        n_antennas = len(mcast_groups) // 2
+        for n, grp in enumerate(mcast_groups[1:]):
+            if grp == mcast_groups[0]:
+                # MK is not using a full power-of-2 size array, there are dummies from here onwards.
+                print("Dummy antenna found starting at ", (n + 1) // 2)
+                n_antennas = (n + 1) // 2
+                break
+        digitiser_address = mcast_groups[0].split("+")[0]
```

Comment:
Not something that needs to get done now, but if you ever wanted to make this more robust/production-ready, it would be best to check that the mcast_groups are all numbered contiguously, have +7 etc. Probably the simplest way to do that would be to reconstruct what sim_correlator is going to think the multicast groups are and check that this matches.

Reply:
I did think about that but I don't know that it's worth being production-ready because this should in principle be CAM's function.

I've watched a few times and it only ever looks like contiguous ranges of addresses are assigned, so we'll probably get away with it this time.